feat: Youtube interactions via Video

marination opened the pull request:
Documentation: https://github.com/frappe/erpnext_documentation/pull/153

- Added Image to Video Doctype (thumbnail if you wish) and Changed Duration field type to Duration
  ![Screenshot 2020-08-02 at 5 44 47 AM](https://user-images.githubusercontent.com/25857446/89112618-03eeb000-d483-11ea-8cb4-5ce95836b4bc.png)

- Youtube Tracking in Video Settings (new)
  Only If tracking is enabled, the **API key** and **Frequency** fields will be visible and mandatory. Frequency is how often the scheduler should auto update the statistics.
 ![Screenshot 2020-09-10 at 6 31 31 PM](https://user-images.githubusercontent.com/25857446/92732539-3a79ed80-f394-11ea-9077-e45ce07804ef.png)


- On enabling tracking in Video Settings , Tracking section will be visible. currently updates realtime on refresh.
  ![Screenshot 2020-08-02 at 5 48 04 AM](https://user-images.githubusercontent.com/25857446/89112627-1b2d9d80-d483-11ea-93b4-cd923739148d.png)

- List View
  ![Screenshot 2020-08-02 at 1 26 31 PM](https://user-images.githubusercontent.com/25857446/89121513-2c0afd00-d4dd-11ea-9189-fcfb4aeb33ee.png)

- Watch Video in Doc
  ![video-2](https://user-images.githubusercontent.com/25857446/89112673-cf2f2880-d483-11ea-96d1-6804cff342dd.gif)

- Interactions Report with Date filters
  ![screencapture-ugh-in-8001-desk-2020-08-02-16_24_19](https://user-images.githubusercontent.com/25857446/89121457-ac7d2e00-d4dc-11ea-8ae3-8daa3bf6fe36.png)
 
**ToDo:**
- [x] Exception Handling and Error Logging

- [x] Job and Frequency

marination committed:
feat: Track Youtube interactions via Video DocType

marination committed:
Merge branch 'develop' of https://github.com/frappe/erpnext into youtube-analytics

marination committed:
chore: Added Interactions Report and behaviour fixes

- Youtube Interactions Report with Chart and Summary
- Statistics change in doc on refresh and get updated in db as well

marination committed:
fix: Codacy

marination committed:
chore: Error Logging and exception hnadling on connection failure

gavindsouza commented on erpnext/utilities/doctype/video/video.py:
```diff
@@ -3,8 +3,51 @@
+				comment_count = %(comment_count)s
+			WHERE name = {0}""".format(frappe.db.escape(docname)), stats) #nosec
+		frappe.db.commit()
+	except:
+		message = "Please make sure you are connected to the Internet"
+		frappe.log_error(message + "\n\n" + frappe.get_traceback(), "Failed to Update YouTube Statistics for Video: {0}".format(docname))
```
```suggestion
	except Exception:
		frappe.log_error("Failed to Update YouTube Statistics for Video: {0}\n\n".format(docname) +  frappe.get_traceback()) 
```
The message variable seems unnecessary and possibly misleading.

eg:

```
Please make sure you are connected to the Internet

Traceback (most recent call last):
  File "/Users/gavin/Desktop/frappe-bench-dev/apps/erpnext/erpnext/utilities/doctype/video/video.py", line 30, in get_video_stats
    video = api.get_video_by_id(video_id=youtube_id)
  File "/Users/gavin/Desktop/frappe-bench-dev/env/lib/python3.7/site-packages/pyyoutube/api.py", line 819, in get_video_by_id
    data = self._parse_response(resp)
  File "/Users/gavin/Desktop/frappe-bench-dev/env/lib/python3.7/site-packages/pyyoutube/api.py", line 290, in _parse_response
    raise PyYouTubeException(response)
pyyoutube.error.PyYouTubeException: YouTubeException(status_code=400,message=API key not valid. Please pass a valid API key.)
```

gavindsouza commented on erpnext/utilities/doctype/video/video.json:
```diff
@@ -15,7 +15,14 @@
-  "description"
+  "description",
+  "image",
+  "youtube_tracking_section",
+  "like_count",
+  "view_count",
+  "col_break",
+  "dislike_count",
+  "comment_count"
```
Can we have these above the description? Seems like it would do better up there instead.

gavindsouza commented on erpnext/utilities/doctype/video/video.js:
```diff
@@ -2,7 +2,30 @@
+	},
+
+	get_video_stats: (frm) => {
+		const expression = '(?:youtube.com/(?:[^/]+/.+/|(?:v|e(?:mbed)?)/|.*[?&]v=)|youtu.be/)([^"&?\\s]{11})';
+		var youtube_id = frm.doc.url.match(expression)[1];
+
+		frappe.call({
+			method: "erpnext.utilities.doctype.video.video.get_video_stats",
+			args: {
+				docname: frm.doc.name,
+				youtube_id: youtube_id
+			}
+		});
+	}
```
Can we handle the video ID extraction on the server-side? Since the doc has the URL, the only args we need to pass is the docname.

gavindsouza commented on erpnext/utilities/doctype/video_settings/video_settings.py:
```diff
@@ -0,0 +1,10 @@
+from frappe.model.document import Document
+
+class VideoSettings(Document):
+	pass
```
Add a check that validates the API Key too.

gavindsouza commented on erpnext/utilities/doctype/video/video.py:
```diff
@@ -3,8 +3,51 @@
 	pass
+
+@frappe.whitelist()
+def get_video_stats(docname, youtube_id, update=True):
```
**Suggestion:**

Move this inside the Video controller under `def get_stats(self): restofyourlogic` you can then use `frm.call` from the form instead.

gavindsouza commented on erpnext/utilities/doctype/video/video.js:
```diff
@@ -2,7 +2,30 @@
 frappe.ui.form.on('Video', {
-	// refresh: function(frm) {
+	refresh: function (frm) {
+		if (frm.doc.provider === "YouTube") {
+			frappe.db.get_single_value("Video Settings", "enable_youtube_tracking").then(value => {
+				if (value) {
+					frm.events.get_video_stats(frm);
+				} else {
+					frm.set_df_property('youtube_tracking_section', 'hidden', true);
+				}
```
Instead of refreshing on every load via Form view _(which could be problematic IMO in case a lot of users view this doc frequently)_, can we add a refresh stats frequency in Video Settings to run via scheduler?

marination committed:
chore: Key Validation and restructure

- Added API Key validation in Video Settings
- Moved Statistics section above description in Video
- Moved id retrieving from URL to py side
- Removed js side calls, added section toggling in js
- Restructured Code

marination committed:
feat: Added Scheduler Job to auto update statistics

- Added frequency in Video Settings
- Cron job to check half hourly job
- Hourly job to run job based on frequency
- Patch to set youtube id in video for ease in computing

marination committed:
Merge branch 'develop' into youtube-analytics

marination committed:
fix: Missing import and typo

marination committed:
fix: Patch reload doctype

marination committed:
Merge branch 'develop' into youtube-analytics

gavindsouza committed:
feat: Add Desk Page for Utilities for Video group

gavindsouza committed:
fix: Commonified function and updated scheduler events

Co-authored-by: marination <maricadsouza221197@gmail.com>

marination committed:
fix: Make set_youtube_statistics reusable and job logic fix

- Make sure job runs only once every hour
- set_youtube_statistics should be usable in and outside class

marination committed:
fix: Separate function to get bulk and single youtube stats

marination committed:
Merge branch 'develop' into youtube-analytics

gavindsouza committed:
Merge branch 'develop' into youtube-analytics

marination committed:
chore: Added video settings to List View Menu
